src/main/java/com/epam/aidial/core/upstream/UpstreamRoute.java

```diff
@@ -1,76 +1,130 @@
+    private final int maxUpstreamsToUse;
 
     /**
-     * @return the number of upstreams which returned any http response.
+     * Current upstream state
      */
-    public int used() {
-        return next;
+    @Nullable
+    private UpstreamState upstreamState;
+    private int used;
+
+    public UpstreamRoute(LoadBalancer<UpstreamState> balancer, int maxUpstreamsToUse) {
+        this.balancer = balancer;
+        this.maxUpstreamsToUse = maxUpstreamsToUse;
+        this.upstreamState = balancer.next();
+        this.used = 1;
     }
 
     /**
-     * @return the number of retries due to connection errors.
+     * @return the number of used upstreams.
      */
-    public int retries() {
-        return Math.min(retries, maxRetries);
+    public int used() {
+        return used;
     }
 
-    @Override
-    public boolean hasNext() {
-        return next < upstreams.size() && retries <= maxRetries;
+    /**
+     * Checks if upstream present (not null) and usage does not exceed max value
+     *
+     * @return true if upstream available, false otherwise
+     */
+    public boolean available() {
+        return upstreamState != null && used <= maxUpstreamsToUse;
     }
 
     /**
-     * @return next endpoint to route to.
+     *  Retrieves next available upstream from load balancer; also increase usage count
+     *
+     * @return next upstream from load balancer
      */
-    @Override
+    @Nullable
     public Upstream next() {
-        if (hasNext()) {
-            prev = next++;
-            int index = (offset + prev) % upstreams.size();
-            upstream = upstreams.get(index);
-            return upstream;
+        // if max attempts reached - do not call balancer
+        if (used + 1 > maxUpstreamsToUse) {
+            this.upstreamState = null;
+            return null;
         }
-
-        return null;
+        used++;
+        UpstreamState upstreamState = balancer.next();
+        this.upstreamState = upstreamState;
+        return upstreamState == null ? null : upstreamState.getUpstream();
     }
 
     /**
-     * @return current endpoint to route to.
+     * @return get current upstream. null if no upstream available
      */
+    @Nullable
     public Upstream get() {
-        Objects.requireNonNull(upstream);
-        return upstream;
+        return upstreamState == null ? null : upstreamState.getUpstream();
     }
 
     /**
-     * Retry the current endpoint because some error happened while sending a request.
+     * Fail current upstream due to error
+     *
+     * @param status - response http status; typically, 5xx or 429
+     * @param retryAfterSeconds - the amount of seconds after which upstream should be available; if status 5xx this value ignored
      */
-    public void retry() {
-        if (prev < next && retries <= maxRetries) {
-            retries++;
-            next = prev;
+    public void fail(HttpStatus status, long retryAfterSeconds) {
+        if (upstreamState == null) {
```

Comment:
`if (upsteamState != null)` is much better flow with only one jump

Reply:
fixed